// @carbon/react/lib/components/MenuButton/index.js
/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var PropTypes = require('prop-types');
var cx = require('classnames');
var iconsReact = require('@carbon/icons-react');
var Button = require('../Button/Button.js');
require('../Button/Button.Skeleton.js');
var Menu = require('../Menu/Menu.js');
require('../Menu/MenuItem.js');
var useAttachedMenu = require('../../internal/useAttachedMenu.js');
var useId = require('../../internal/useId.js');
var usePrefix = require('../../internal/usePrefix.js');
var react = require('@floating-ui/react');
var index = require('../FeatureFlags/index.js');
var mergeRefs = require('../../tools/mergeRefs.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var cx__default = /*#__PURE__*/_interopDefaultLegacy(cx);

const validButtonKinds = ['primary', 'tertiary', 'ghost'];
const defaultButtonKind = 'primary';
const MenuButton = /*#__PURE__*/React.forwardRef(function MenuButton(_ref, forwardRef) {
  let {
    children,
    className,
    disabled,
    kind = defaultButtonKind,
    label,
    size = 'lg',
    menuAlignment = 'bottom',
    tabIndex = 0,
    menuTarget,
    ...rest
  } = _ref;
  // feature flag utilized to separate out only the dynamic styles from @floating-ui
  // flag is turned on when collision detection (ie. flip, hide) logic is not desired
  const enableOnlyFloatingStyles = index.useFeatureFlag('enable-v12-dynamic-floating-styles');
  const id = useId.useId('MenuButton');
  const prefix = usePrefix.usePrefix();
  const triggerRef = React.useRef(null);
  let middlewares = [];
  if (!enableOnlyFloatingStyles) {
    middlewares = [react.flip({
      crossAxis: false
    })];
  }
  if (menuAlignment === 'bottom' || menuAlignment === 'top') {
    middlewares.push(react.size({
      apply(_ref2) {
        let {
          rects,
          elements
        } = _ref2;
        Object.assign(elements.floating.style, {
          width: `${rects.reference.width}px`
        });
      }
    }));
  }
  const {
    refs,
    floatingStyles,
    placement,
    middlewareData
  } = react.useFloating({
    placement: menuAlignment,
    // The floating element is positioned relative to its nearest
    // containing block (usually the viewport). It will in many cases also
    // “break” the floating element out of a clipping ancestor.
    // https://floating-ui.com/docs/misc#clipping
    strategy: 'fixed',
    // Middleware order matters, arrow should be last
    middleware: middlewares,
    whileElementsMounted: react.autoUpdate
  });
  const ref = mergeRefs["default"](forwardRef, triggerRef);
  const {
    open,
    handleClick: hookOnClick,
    handleMousedown,
    handleClose
  } = useAttachedMenu.useAttachedMenu(triggerRef);
  React.useLayoutEffect(() => {
    Object.keys(floatingStyles).forEach(style => {
      if (refs.floating.current) {
        refs.floating.current.style[style] = floatingStyles[style];
      }
    });
  }, [floatingStyles, refs.floating, middlewareData, placement, open]);
  function handleClick() {
    if (triggerRef.current) {
      hookOnClick();
    }
  }
  const containerClasses = cx__default["default"](`${prefix}--menu-button__container`, className);
  const triggerClasses = cx__default["default"](`${prefix}--menu-button__trigger`, {
    [`${prefix}--menu-button__trigger--open`]: open
  });
  const menuClasses = cx__default["default"](`${prefix}--menu-button__${menuAlignment}`);
  return /*#__PURE__*/React__default["default"].createElement("div", _rollupPluginBabelHelpers["extends"]({}, rest, {
    ref: ref,
    "aria-owns": open ? id : undefined,
    className: containerClasses
  }), /*#__PURE__*/React__default["default"].createElement(Button["default"], {
    ref: refs.setReference,
    className: triggerClasses,
    size: size,
    tabIndex: tabIndex,
    kind: kind,
    renderIcon: iconsReact.ChevronDown,
    disabled: disabled,
    "aria-haspopup": true,
    "aria-expanded": open,
    onClick: handleClick,
    onMouseDown: handleMousedown,
    "aria-controls": open ? id : undefined
  }, label), /*#__PURE__*/React__default["default"].createElement(Menu.Menu, {
    containerRef: triggerRef,
    menuAlignment: menuAlignment,
    className: menuClasses,
    ref: refs.setFloating,
    id: id,
    legacyAutoalign: false,
    label: label,
    mode: "basic",
    size: size,
    open: open,
    onClose: handleClose,
    target: menuTarget
  }, children));
});
MenuButton.propTypes = {
  /**
   * A collection of MenuItems to be rendered as actions for this MenuButton.
   */
  children: PropTypes__default["default"].node.isRequired,
  /**
   * Additional CSS class names.
   */
  className: PropTypes__default["default"].string,
  /**
   * Specify whether the MenuButton should be disabled, or not.
   */
  disabled: PropTypes__default["default"].bool,
  /**
   * Specify the type of button to be used as the base for the trigger button.
   */
  // @ts-ignore-next-line -- avoid spurious (?) TS2322 error
  kind: PropTypes__default["default"].oneOf(validButtonKinds),
  /**
   * Provide the label to be rendered on the trigger button.
   */
  label: PropTypes__default["default"].string.isRequired,
  /**
   * Experimental property. Specify how the menu should align with the button element
   */
  menuAlignment: PropTypes__default["default"].oneOf(['top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end']),
  /**
   * Specify the size of the button and menu.
   */
  // @ts-ignore-next-line -- avoid spurious (?) TS2322 error
  size: PropTypes__default["default"].oneOf(['sm', 'md', 'lg']),
  /**
   * Specify the tabIndex of the button.
   */
  // @ts-ignore-next-line -- avoid spurious (?) TS2322 error
  tabIndex: PropTypes__default["default"].number,
  /**
   * Specify a DOM node where the Menu should be rendered in. Defaults to document.body.
   */

  menuTarget: PropTypes__default["default"].instanceOf(typeof Element !== 'undefined' ? Element : Object)
};

exports.MenuButton = MenuButton;
